test/metabase/api/model_index_test.clj

```diff
@@ -0,0 +1,87 @@
+(ns metabase.api.model-index-test
+  (:require [clojure.core.async :as a]
+            [clojure.test :refer :all]
+            [metabase.models.card :refer [Card]]
+            [metabase.query-processor.async :as qp.async]
+            [metabase.test :as mt]
+            [toucan2.util :as u]))
+
+(defn- result-metadata-for-query [query]
+  (first
+   (a/alts!!
+    [(qp.async/result-metadata-for-query-async query)
+     (a/timeout 2000)])))
+
+(deftest full-lifecycle-test
+  (mt/dataset sample-dataset
+    (let [query     (mt/mbql-query products)
+          pk_ref    (mt/$ids $products.id)
+          value_ref (mt/$ids $products.title)]
+      (mt/with-temp* [Card [model {:dataset         true
+                                   :name            "model index test"
+                                   :dataset_query   query
+                                   :result_metadata (result-metadata-for-query query)}]]
```

Comment:
```suggestion
      (mt/with-temp* [Card [model (merge (mt/card-with-source-metadata-for-query query)
       {:dataset true
        :name    "model index test"})]]
```

this does the same thing

Reply:
also, I think we should start using `t2.with-temp/with-temp` to make the transition easier